File: langhuan-service/src/main/java/com/langhuan/utils/rag/splitter/LlmTextSplitter.java
package com.langhuan.utils.rag.splitter;

import com.langhuan.serviceai.ChatGeneralAssistanceService;

import cn.hutool.json.JSONArray;
import cn.hutool.json.JSONUtil;
import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.List;

@Slf4j
public class LlmTextSplitter {

    private final int windowSize;
    private final String modelName;
    private final ChatGeneralAssistanceService chatGeneralAssistanceService;

    public LlmTextSplitter(int windowSize, String modelName, ChatGeneralAssistanceService chatGeneralAssistanceService) {
        this.windowSize = windowSize;
        this.modelName = modelName;
        this.chatGeneralAssistanceService = chatGeneralAssistanceService;
    }

    public List<String> apply(String text) {
        List<String> documents = new ArrayList<>();
        int length = text.length();
        for (int i = 0; i < length; i += windowSize) {
            String chunk = text.substring(i, Math.min(length, i + windowSize));
            log.info("llmOutString:{}", chunk);
            String llmOutString = this.chatGeneralAssistanceService.llmTextSplitter(modelName, chunk);
            llmOutString = llmOutString.replace("```json", "");
            llmOutString = llmOutString.replace("```", "");
            JSONArray content = JSONUtil.parseObj(llmOutString).getJSONArray("content");
            List<String> list = content.stream()
                    .map(Object::toString)
                    .toList();
            documents.addAll(list);
        }
        return documents;
    }


}

File: langhuan-server/src/main/kotlin/com/langhuan/utils/rag/splitter/LlmTextSplitter.kt
package com.langhuan.utils.rag.splitter

import cn.hutool.json.JSONArray
import cn.hutool.json.JSONUtil
import com.langhuan.serviceai.ChatGeneralAssistanceService
import org.slf4j.LoggerFactory

class LlmTextSplitter(
    private val windowSize: Int,
    private val modelName: String,
    private val chatGeneralAssistanceService: ChatGeneralAssistanceService
) : TextSplitter {

    companion object {
        private val log = LoggerFactory.getLogger(LlmTextSplitter::class.java)
    }

    override fun apply(text: String): List<String> {
        val documents = mutableListOf<String>()
        val length = text.length
        for (i in 0 until length step windowSize) {
            val chunk = text.substring(i, minOf(length, i + windowSize))
            log.info("llmOutString:{}", chunk)
            var llmOutString = chatGeneralAssistanceService.llmTextSplitter(modelName, chunk)
            if (llmOutString != null) {
                llmOutString = llmOutString.replace("```json", "")
            }
            if (llmOutString != null) {
                llmOutString = llmOutString.replace("```", "")
            }
            val content = JSONUtil.parseObj(llmOutString).getJSONArray("content")
            val list = content.stream()
                    .map { it.toString() }
                    .toList()
            documents.addAll(list)
        }
        return documents
    }
}
